catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,129 @@
+import java.util.Map;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.federation.FederationException;
+import ddf.catalog.operation.SourceResponse;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+
+public class QueryResultPaginator {
+
+    private final CatalogFramework catalogFramework;
+
+    private final QueryRequestImpl queryRequest;
+
+    private SourceResponse response = null;
+
+    private int ceilingIndex;
+
+    private int endIndex;
+
+    private int maxResults;
+
+    private int indexOffset;
+
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryRequestImpl queryRequest) {
+        this.catalogFramework = catalogFramework;
+        this.queryRequest = queryRequest;
+
+        // maxResults == total number of results to return to client
+        maxResults = queryRequest.getQuery().getPageSize();
+
+        indexOffset = 1;
+        endIndex = 100000000;
+        ceilingIndex = 100000000;
+    }
+
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryRequestImpl queryRequest,
+            int indexOffset, int endIndex, int maxResults, int ceilingIndex) throws Exception {
+
+        this.catalogFramework = catalogFramework;
+        this.queryRequest = queryRequest;
+        this.indexOffset = indexOffset;
+        this.endIndex = endIndex;
+        this.maxResults = maxResults;
+        this.ceilingIndex = ceilingIndex;
+
+        if (indexOffset < 0 || indexOffset > ceilingIndex || indexOffset > endIndex) {
+            throw new Exception();
+        }
+    }
+
+    // TODO: 6/14/17 Determine how this is used and best way to check
+    //    public boolean hasNext() {
+    //        boolean hasNext;
+    //
+    //        if (maxResults > 0) {
+    //            hasNext = true;
+    //        } else {
+    //            hasNext = false;
+    //        }
+    //
+    //        return hasNext;
+    //    }
+
+    public List<Result> nextPage() {
+        List<Result> resultList = new ArrayList<>();
+
+        final int originalIndexOffset = indexOffset;
+
+        while (indexOffset < endIndex && resultList.size() < maxResults
+                && indexOffset < ceilingIndex) {
+
+            resultList.addAll(queryCatalogFramework());
+            resultList = trimListToMaxResults(resultList,
+                    maxResults,
+                    ceilingIndex,
+                    originalIndexOffset);
+
+            indexOffset = originalIndexOffset + resultList.size();
+        }
+        return resultList;
+    }
+
+    private List<Result> queryCatalogFramework() {
+        try {
+            Map<String, Serializable> props = new HashMap<>();
+            // Avoid caching all iteratorResults while dumping with native query mode
+            props.put("mode", "native");
+            response = catalogFramework.query(queryRequest);
```

Comment:
❗️ We need to update the query's start index before calling the catalog otherwise we will always get the same results over and over again. We also need to make sure our unit tests catch this.